### PR TITLE
ci.yml: Upgrade GitHub Actions

### PR DESCRIPTION
**Pull request by cclauss**
* https://github.com/actions/checkout/releases
* https://github.com/actions/setup-python/releases
* https://github.com/actions/upload-artifact/releases
* https://github.com/deadsnakes/action/releases

This could be automated via dependabot or other similar tools.

### CODE REVIEW

**Comment by mattip** [comment]
The PR is failing

**Comment by cclauss** [comment]
Yes.  I think we will need to keep using deadsnakes < v3.0... https://github.com/deadsnakes/action/releases/tag/v3.0.0

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -67,10 +67,10 @@ jobs:
         python-version: ['3.12', '3.11', '3.10', '3.9', '3.8']
 
     steps:
-      - uses: actions/checkout@v3
+      - uses: actions/checkout@v4
 
       - name: Set up Python from deadsnakes
-        uses: deadsnakes/action@v2.1.1
+        uses: deadsnakes/action@v3.1.0
```
Comment (cclauss) [inline .github/workflows/ci.yml:73]:
```suggestion
        uses: deadsnakes/action@v2.1.1  # Upgrading will cause test failures.
```

**Review by fangerer — APPROVED** [review]
lgtm, thanks

**Pull request merged**